src/store/api/worldUsers.ts

```diff
@@ -0,0 +1,231 @@
+      }),
+      async onCacheEntryAdded(
+        { relatedLocationIds },
+        { updateCachedData, cacheEntryRemoved }
+      ) {
+        // Used to hold the changes queued from the snapshot listener so that we can proces them in batches
+        const queuedChanges: firebase.firestore.DocumentChange<firebase.firestore.DocumentData>[] = [];
+
+        const processQueuedChanges = () => {
+          if (queuedChanges.length === 0) return;
+
+          updateCachedData((draft) => {
+            // We use splice here to remove all elements from the array and return them for processing
+            queuedChanges.splice(0).forEach(processUserDocChange(draft));
+          });
+        };
+
+        // TODO: move this interval into a proper const/config value somewhere
+        const processQueuedChangesIntervalId = setInterval(
+          processQueuedChanges,
+          5000
+        );
+
+        const worldUsersQuery = firebase
+          .firestore()
+          .collection("users")
+          .where("enteredVenueIds", "array-contains-any", relatedLocationIds);
+
+        const unsubscribeListener = worldUsersQuery.onSnapshot((snapshot) => {
+          // TODO: implement this properly + cleanup this placeholder stuff
+          const added = snapshot
+            .docChanges()
+            .filter((docChange) => docChange.type === "added").length;
+
+          const removed = snapshot
+            .docChanges()
+            .filter((docChange) => docChange.type === "removed").length;
+
+          const modified = snapshot
+            .docChanges()
+            .filter((docChange) => docChange.type === "modified").length;
+
+          console.log(
+            "worldUsersApi::worldUsers::snapshot",
+            "\n  snapshot.size=",
+            snapshot.size,
+            "\n  snapshot.docs.length = ",
+            snapshot.docs.length,
+            "\n  snapshot.docChanges().length",
+            snapshot.docChanges().length,
+            "\n  snapshot.docChanges() added:",
+            added,
+            "\n  snapshot.docChanges() removed:",
+            removed,
+            "\n  snapshot.docChanges() modified:",
+            modified,
+            "\n  queuedChanges.length",
+            queuedChanges.length
+          );
+
+          snapshot.docChanges().forEach((change) => {
+            // TODO: check if this document relates to the current user, if so, update it immediately
+            //   Note that we will need to pass their userId in and/or read it from somehwere to make this work..
+            //   probably just the queryArgs for this whole api listener
+            //
+            // const currentUserId = "TODO";
+            //
+            // if (change.doc.id === currentUserId) {
+            //   updateCachedData((draft) => {
+            //     processUserDocChange(draft)(change);
+            //   });
+            // } else {
+            queuedChanges.push(change);
+            // }
+          });
+        });
```

Comment:
[note] Here is where we handle the main snapshot processing, with some placeholder boilerplate for ensuring that we immediately process any actions related to the current user so that we don't need to hack around cache invalidation/end up with stale data when taking some actions on the platform (eg. taking a seat)

Basically, for each change from `snapshot.docChanges()` (usually seems to only be a single change per snapshot for user updates) we just `.push` that change into the mutable `queuedChanges` array, which will then be processed later by the `processQueuedChanges` we defined earlier.

When implemented, if the change is to the current user's document, then we will just immediately process it with `processUserDocChange` within Immer's `updateCachedData` instead of adding it to `queuedChanges`.